pkg/reconciler/ingress/resources/gateway.go

```diff
@@ -83,9 +89,80 @@ func SortServers(servers []v1alpha3.Server) []v1alpha3.Server {
 	return servers
 }
 
-// MakeServers creates the expected Gateway `Servers` based on the given
-// ClusterIngress.
-func MakeServers(ia v1alpha1.IngressAccessor, gatewayServiceNamespace string, originSecrets map[string]*corev1.Secret) ([]v1alpha3.Server, error) {
+// MakeIngressGateways creates Gateways for a given IngressAccessor.
+func MakeIngressGateways(ctx context.Context, ia v1alpha1.IngressAccessor, originSecrets map[string]*corev1.Secret, svcLister corev1listers.ServiceLister) ([]*v1alpha3.Gateway, error) {
+	gatewayServices, err := getGatewayServices(ctx, svcLister)
+	if err != nil {
+		return nil, err
+	}
+	gateways := make([]*v1alpha3.Gateway, len(gatewayServices))
+	for i, gatewayService := range gatewayServices {
+		gateway, err := makeIngressGateway(ctx, ia, originSecrets, gatewayService.Spec.Selector, gatewayService)
+		if err != nil {
+			return nil, err
+		}
+		gateways[i] = gateway
+	}
+	return gateways, nil
+}
+
+func makeIngressGateway(ctx context.Context, ia v1alpha1.IngressAccessor, originSecrets map[string]*corev1.Secret, selector map[string]string, gatewayService *corev1.Service) (*v1alpha3.Gateway, error) {
+	ns := ia.GetNamespace()
+	if len(ns) == 0 {
+		ns = system.Namespace()
+	}
+	servers, err := MakeTLSServers(ia, gatewayService.Namespace, originSecrets)
+	if err != nil {
+		return nil, err
+	}
+	hosts := sets.String{}
+	for _, rule := range ia.GetSpec().Rules {
+		hosts.Insert(rule.Hosts...)
+	}
+	servers = append(servers, *MakeHTTPServer(config.FromContext(ctx).Network.HTTPProtocol, hosts.List()))
+	return &v1alpha3.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            GatewayName(ia, gatewayService),
+			Namespace:       ns,
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(ia)},
+			Labels: map[string]string{
+				// We need this label to find out all of Gateways of a given Ingress.
+				networking.IngressLabelKey: ia.GetName(),
+			},
+		},
+		Spec: v1alpha3.GatewaySpec{
+			Selector: selector,
+			Servers:  servers,
+		},
+	}, nil
+}
+
+func getGatewayServices(ctx context.Context, svcLister corev1listers.ServiceLister) ([]*corev1.Service, error) {
+	ingressSvcMetas, err := getIngressGatewaySvcNameNamespaces(ctx)
+	if err != nil {
+		return nil, err
+	}
+	services := []*corev1.Service{}
```

Comment:
well same preallocation here can be done, right?

Reply:
Done